opm/simulators/wells/GasLiftStage2_impl.hpp

```diff
@@ -0,0 +1,1027 @@
+GasLiftStage2<TypeTag>::
+deleteDecGradItem_(const std::string &name)
+{
+    return deleteGrad_(name, /*increase=*/false);
+}
+
+template<typename TypeTag>
+typename GasLiftStage2<TypeTag>::GradInfo
+GasLiftStage2<TypeTag>::
+deleteIncGradItem_(const std::string &name)
+{
+    return deleteGrad_(name, /*increase=*/true);
+}
+
+template<typename TypeTag>
+void
+GasLiftStage2<TypeTag>::
+displayWarning_(const std::string &msg, const std::string &group_name)
+{
+    const std::string message = fmt::format(
+        "GAS LIFT OPTIMIZATION (STAGE2), GROUP: {} : {}", group_name, msg);
+    this->deferred_logger_.warning("WARNING", message);
+}
+
+template<typename TypeTag>
+void
+GasLiftStage2<TypeTag>::
+displayWarning_(const std::string &msg)
+{
+    const std::string message = fmt::format(
+        "GAS LIFT OPTIMIZATION (STAGE2) : {}", msg);
+    this->deferred_logger_.warning("WARNING", message);
+}
+
+template<typename TypeTag>
+void
+GasLiftStage2<TypeTag>::
+displayDebugMessage_(const std::string &msg)
+{
+    if (this->debug_) {
+        const std::string message = fmt::format(
+            "  GLIFT2 (DEBUG) : {}", msg);
+        this->deferred_logger_.info(message);
+    }
+}
+
+template<typename TypeTag>
+void
+GasLiftStage2<TypeTag>::
+displayDebugMessage2B_(const std::string &msg)
+{
+    if (this->debug_) {
+        const std::string message = fmt::format(
+            "Stage 2B : {}", msg);
+        displayDebugMessage_(message);
+    }
+}
+template<typename TypeTag>
+void
+GasLiftStage2<TypeTag>::
+displayDebugMessage_(const std::string &msg, const std::string &group_name)
+{
+    if (this->debug_) {
+        const std::string message = fmt::format(
+            "Group {} : {}", group_name, msg);
+        displayDebugMessage_(message);
+    }
+}
+
+template<typename TypeTag>
+std::tuple<double, double, double>
+GasLiftStage2<TypeTag>::
+getCurrentGroupRates_(const Opm::Group &group)
+{
+    if (this->debug_) {
+        auto [oil_rate, gas_rate, alq] = getCurrentGroupRatesRecursive_(group);
+        const std::string msg = fmt::format(
+            "Current group rates for {} : oil: {}, gas: {}, alq: {}",
+            group.name(), oil_rate, gas_rate, alq);
+        displayDebugMessage2B_(msg);
+        return {oil_rate, gas_rate, alq};
+    }
+    else {
+        return getCurrentGroupRatesRecursive_(group);
+    }
+}
+
+
+template<typename TypeTag>
+std::tuple<double, double, double>
+GasLiftStage2<TypeTag>::
+getCurrentGroupRatesRecursive_(const Opm::Group &group)
+{
+    double oil_rate = 0.0;
+    double gas_rate = 0.0;
+    double alq = 0.0;
+    // NOTE: A group can either contain wells or groups, but not both
+    if (group.wellgroup()) {
+        for (const std::string& well_name : group.wells()) {
+            auto [sw_oil_rate, sw_gas_rate, sw_alq] =
+                getCurrentWellRates_(well_name, group.name());
+            oil_rate += sw_oil_rate;
+            gas_rate += sw_gas_rate;
+            alq += sw_alq;
```

Comment:
In the spirit of #3095, please double check the group efficiency is also applied for the top-most group.
If this is based on code before the merge, chances are that it is not